tests/e2e/runtime_reconfiguration_test.go

```diff
@@ -0,0 +1,201 @@
+	}{
+		{
+			name:        "grow cluster size from 1 to 3",
+			clusterSize: 1,
+		},
+		{
+			name:        "grow cluster size from 3 to 5",
+			clusterSize: 3,
+		},
+		{
+			name:        "grow cluster size from 1 to 3 with learner",
+			clusterSize: 1,
+			asLearner:   true,
+		},
+		{
+			name:        "grow cluster size from 3 to 5 with learner",
+			clusterSize: 3,
+			asLearner:   true,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			epc, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithClusterSize(tc.clusterSize))
+			require.NoError(t, err)
+			require.NoError(t, epc.Procs[0].Etcdctl().Health(ctx))
+			defer func() {
+				err := epc.Close()
+				require.NoError(t, err, "failed to close etcd cluster: %v", err)
+			}()
+
+			for i := 0; i < 2; i++ {
+				time.Sleep(etcdserver.HealthInterval)
+				if !tc.asLearner {
+					addMember(ctx, t, epc)
+				} else {
+					addMemberAsLearner(ctx, t, epc)
+				}
+			}
+		})
+	}
+}
+
+func TestRuntimeReconfigDecreaseClusterSize(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	tcs := []struct {
+		name        string
+		clusterSize int
+		asLearner   bool
+	}{
+		{
+			name:        "decrease cluster size from 3 to 1",
+			clusterSize: 3,
+		},
+		{
+			name:        "decrease cluster size from 5 to 3",
+			clusterSize: 5,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			epc, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithClusterSize(tc.clusterSize))
+			require.NoError(t, err)
+			require.NoError(t, epc.Procs[0].Etcdctl().Health(ctx))
+			defer func() {
+				err := epc.Close()
+				require.NoError(t, err, "failed to close etcd cluster: %v", err)
+			}()
+
+			for i := 0; i < 2; i++ {
+				time.Sleep(etcdserver.HealthInterval)
+				removeMember(ctx, t, epc)
+			}
+		})
+	}
+}
+
+func TestRuntimeReconfigRollingUpgrade(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	tcs := []struct {
+		name        string
+		withLearner bool
+	}{
+		{
+			name:        "with learner",
+			withLearner: true,
+		},
+		{
+			name:        "without learner",
+			withLearner: false,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			epc, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithClusterSize(3))
+			require.NoError(t, err)
+			require.NoError(t, epc.Procs[0].Etcdctl().Health(ctx))
+			defer func() {
+				err := epc.Close()
+				require.NoError(t, err, "failed to close etcd cluster: %v", err)
+			}()
+
+			for i := 0; i < 2; i++ {
+				time.Sleep(etcdserver.HealthInterval)
+				removeMember(ctx, t, epc)
+				epc.WaitLeader(t)
+				// if we do not wait for leader, without the fix of notify raft Advance,
+				// have to wait 1 sec to pass the test stably.
```

Comment:
Is there a github issue we can link to for the notify raft advance fix?

Reply:
Those will be fixed https://github.com/etcd-io/etcd/issues/15528, https://github.com/etcd-io/etcd/issues/16078, https://github.com/etcd-io/etcd/issues/14040, https://github.com/etcd-io/etcd/issues/13212, https://github.com/etcd-io/etcd/issues/12983